### PR TITLE
Design and multi-residency support

### PR DESCRIPTION
**Pull request by jczaja**

### CODE REVIEW

**Comment by jczaja** [comment]
@sfraczek, @tpatejko Please review

File: src/pl.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        while is_success == false {
+            converted_date = converted_date
+                .checked_sub_signed(chrono::Duration::days(1))
+                .expect("Error traversing date");
+
+            let exchange_rate_url: String = base_exchange_rate_url.to_string()
+                + &format!("{}", converted_date.format("%Y-%m-%d"))
+                + "/?format=json";
+
+            let body = client.get(&(exchange_rate_url)).send();
```
Comment (sfraczek) [inline src/pl.rs:69]: I personally would remove extra parentheses

File: src/pl.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            is_success = actual_body.status().is_success();
+            if is_success == true {
+                log::info!("RESPONSE {:#?}", actual_body);
+
+                let nbp_response = actual_body
+                    .json::<NBPResponse<ExchangeRate>>()
+                    .expect("Error converting response to JSON");
+                log::info!("body of exchange_rate = {:#?}", nbp_response);
+                exchange_rate = nbp_response.rates[0].mid;
+                exchange_rate_date = format!("{}", converted_date.format("%Y-%m-%d"));
```
Comment (sfraczek) [inline src/pl.rs:83]: Maybe this converted_date could be converted to string once because this is happening second time here.

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pub trait Residency {
+    fn get_exchange_rate(&self, transaction_date: &str) -> Result<(String, f32), String>;
```
Comment (sfraczek) [inline src/lib.rs:2]: You could consider changing return type from Result to simpler one until you add returning of errors just for simplicity.

File: src/main.rs [hunk 4]
```diff
@@ -199,21 +120,52 @@ fn compute_tax(transactions: Vec<Transaction>) -> (f32, f32) {
...
+            Arg::with_name("pdf documents")
+                .help("Brokerage statement PDF files")
+                .multiple(true),
+        )
+        .get_matches();
+
+    let residency = matches
+        .value_of("residence")
+        .expect("error getting residence value");
+    let rd: Box<dyn etradeTaxReturnHelper::Residency> = match residency {
```
Comment (sfraczek) [inline src/main.rs:142]: I would prefer longer variable name

File: src/main.rs [hunk 5]
```diff
@@ -199,21 +120,52 @@ fn compute_tax(transactions: Vec<Transaction>) -> (f32, f32) {
...
+        "pl" => Box::new(pl::PL {}),
+        "usd" => Box::new(us::US {}),
+        _ => panic!(
+            "{}",
+            &format!("Error: unimplemented residency: {}", residency)
```
[the remark this answers is not in the document]
Comment (jczaja) [reply src/main.rs:147]: I do not panic macro is supporting format. This is why I had to go through format!

**Pull request merged**